src/main/kotlin/com/fasterxml/jackson/module/kotlin/KotlinNamesAnnotationIntrospector.kt

```diff
@@ -18,38 +18,42 @@ import java.util.Locale
 import kotlin.reflect.KClass
 import kotlin.reflect.KFunction
 import kotlin.reflect.KParameter
-import kotlin.reflect.full.companionObject
-import kotlin.reflect.full.declaredFunctions
-import kotlin.reflect.full.memberProperties
-import kotlin.reflect.full.primaryConstructor
+import kotlin.reflect.full.*
 import kotlin.reflect.jvm.internal.KotlinReflectionInternalError
+import kotlin.reflect.jvm.javaGetter
 import kotlin.reflect.jvm.javaType
 import kotlin.reflect.jvm.kotlinFunction
 
 internal class KotlinNamesAnnotationIntrospector(val module: KotlinModule, val cache: ReflectionCache, val ignoredClassesForImplyingJsonCreator: Set<KClass<*>>) : NopAnnotationIntrospector() {
     // since 2.4
-    override fun findImplicitPropertyName(member: AnnotatedMember): String? {
-        if (member is AnnotatedMethod && member.isInlineClass()) {
-            if (member.name.startsWith("get") &&
-                    member.name.contains('-') &&
-                    member.parameterCount == 0) {
-                return member.name.substringAfter("get")
-                    .replaceFirstChar { it.lowercase(Locale.getDefault()) }
-                    .substringBefore('-')
-            } else if (member.name.startsWith("is") &&
-                    member.name.contains('-') &&
-                    member.parameterCount == 0) {
-                return member.name.substringAfter("is")
-                    .replaceFirstChar { it.lowercase(Locale.getDefault()) }
-                    .substringBefore('-')
-            }
-        } else if (member is AnnotatedParameter) {
-            return findKotlinParameterName(member)
-        }
-
-        return null
+    override fun findImplicitPropertyName(member: AnnotatedMember): String? = when (member) {
+        is AnnotatedMethod -> findImplicitPropertyNameFromKotlinPropertyIfNeeded(member)
+        is AnnotatedParameter -> findKotlinParameterName(member)
+        else -> null
     }
 
+    // Since getter for value class (inline class) will be compiled into a different name such as "getFoo-${random}",
+    // use the property name in Kotlin in such a case.
+    private fun findImplicitPropertyNameFromKotlinPropertyIfNeeded(member: AnnotatedMethod): String? = member
+        .takeIf {
+            it.parameterCount == 0 &&
+                    it.name.run { contains("-") && (startsWith("get") || startsWith("is")) }
+        }?.let { _ ->
+            val propertyNameFromGetter = member.name.let {
+                when {
+                    it.startsWith("get") -> it.substringAfter("get")
+                    it.startsWith("is") -> it.substringAfter("is")
+                    else -> throw IllegalStateException("Should not get here.")
+                }
+            }.replaceFirstChar { it.lowercase(Locale.getDefault()) }
```

Comment:
I haven't opened this in an IDE so could be wrong, but does this work if you use `when (member.name)` and drop the `.let {}`?

```kotlin
            val propertyNameFromGetter = when (member.name) {
                    it.startsWith("get") -> it.substringAfter("get")
                    it.startsWith("is") -> it.substringAfter("is")
                    else -> throw IllegalStateException("Should not get here.")
            }.replaceFirstChar { it.lowercase(Locale.getDefault()) }
```

Reply:
I tried, but it seemed that I could only write the following.

```kotlin
val propertyNameFromGetter = when  {
    member.name.startsWith("get") -> member.name.substringAfter("get")
    member.name.startsWith("is") -> member.name.substringAfter("is")
    else -> throw IllegalStateException("Should not get here.")
}.replaceFirstChar { it.lowercase(Locale.getDefault()) }
```

I thought it would be better to use `let` because `member.name` would be repeated many times.

Reply:
Quick note: lower-casing may be incorrect here. Consider case of `getURL()` which should produce, I think, `URL`, as per Java Beans definition. But if code above simply lower-cases the first character, it'd return `uRL`.
(Jackson itself has bit of inconsistency, originally returning `url` for that case, lower-casing all leading upper-case chars; but then adding option for producing `URL`)

Reply:
@cowtowncoder 
In this scene, if the name retrieved from the `getter` is different from the name of the property, the name of the property will be adopted.
In other words, even if the name retrieved from the `getter` is wrong, the name of the property (= expected value) will be adopted as a result.
For this reason, I thought the problem you pointed out would not occur here.

Reply:
I noticed that the pattern where `getter` is named with `@JvmGetter` doesn't cover some areas.
It looks like the property names should not be used in this pattern.

I will revisit this later as I don't have much more time right now.

Reply:
I am not sure exactly how much Kotlin module changes things, but the way basic Jackson property inspection works is that it first gets all "implicit" (and explicit) names of accessors, then joins similarly named entries (by implicit name), then does renaming of grouped accessors.
In this scheme, then, if implicit name from getter does not match those of fields, they would infer different non-related properties.

I have been hoping to change handling to let modules (Kotlin and Scala mostly) customize logic more to essentially "start with fields" and use different merging logic. But right now ways to override things tend to have unexpected side-effects.

For this particular case, also, I suspect that if matching of accessors used case-insensitive handling (not trivial to do as retrofit but something that has been requested and seems useful), mismatch could be avoided. And like you say, could then just use name from field (implicit name or explicit annotation)/